accounting/invoice/utils.py

```diff
@@ -21,27 +21,132 @@
+    file = invoice.upload_to_google_drive(
+        file_path=invoice.pdf_path,
+        target_path=target_path,
+        title=invoice.pdf_filename,
+    )
+    invoice.pdf_path = file['webContentLink']
+    invoice.save()
 
 
-def default_invoice_number():
+# pylint: disable=too-many-arguments,too-many-locals
+def send_email_with_invoice(template, subject, extra_email_context=None, client_usernames=settings.BILLING_CYCLE_USERS,
+                            create_invoice=False, draft_invoice=False, fill_line_items_from_jira=False,
+                            upload_to_google_drive=False, auto_approve=False):
     """
-    Return the invoice number which defaults to `yyyy-mm` for the current year and month.
+    Send an email with an invoice attached, and potentially perform peripheral actions.
+
+    TODO: This function's pretty beefy; break it up by encapsulating logic into more utility functions or a new class.
+
+    :param template: The email template.
+    :param subject: The email subject.
+    :param extra_email_context: Extra kwargs to update the email context with.
+    :param client_usernames: The list of billing cycle users' usernames.
+    :param create_invoice: Whether to create a new invoice based off of a past one.
+    :param draft_invoice: Whether this is a draft invoice or not.
+    :param fill_line_items_from_jira: Whether to fill up the invoice's line items from JIRA worklogs.
+    :param upload_to_google_drive: Whether to upload the resulting invoice PDF to Google Drive.
+                                   If using a draft invoice, it gets uploaded into `invoices-in/draft/{month}`.
     """
-    return timezone.now().strftime("%Y-%m")
+    if extra_email_context is None:
+        extra_email_context = {}
+    now = datetime.datetime.now()
+    past = get_last_day_past_month()
+    past_month_formatted = past.strftime('%B')
+    template = get_template(template)
+    for client_username in client_usernames:
+        client = Account.objects.get(user__username=client_username)
+        providers = [rate.provider for rate in client.client_hourly_rates.filter(active=True)]
+        for provider in providers:
+            try:
+                invoice = models.Invoice.objects.filter(
+                    provider=provider,
+                    client=client,
+                    date__month=past.month if draft_invoice else now.month,
+                ).latest('date')
+            except models.Invoice.DoesNotExist:
+                # This may be the provider's first invoice.
+                invoice = None
+
+            if create_invoice or not invoice:
+                invoice = models.Invoice.objects.create(
+                    number=choices.InvoiceNumberingScheme.increment_value(
+                        invoice.template.numbering_scheme,
+                        invoice.number
+                    ) if invoice else choices.InvoiceNumberingScheme. provider.invoice_template.numbering_scheme,
```

Comment:
@UmanShahzad This looks like an "Undo" gone bad? 😄

Reply:
:man_facepalming: